cmd/kubeadm/app/util/kubeconfig/kubeconfig_test.go

```diff
@@ -345,3 +346,198 @@ func TestHasCredentials(t *testing.T) {
+			}
+			if !reflect.DeepEqual(got1, rt.expected1) {
+				t.Errorf("GetClusterFromKubeConfig() got1 = %v, expected1 %v", got1, rt.expected1)
+			}
+		})
+	}
+}
+
+func TestEnsureAuthenticationInfoAreEmbedded(t *testing.T) {
+	file, dir, err := createtempfile(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer os.RemoveAll(dir)
+	defer file.Close()
+
+	tests := []struct {
+		name    string
+		config  *clientcmdapi.Config
+		wantErr bool
+	}{
+		{
+			name:    "invalid authInfo: no authInfo",
+			config:  nil,
+			wantErr: true,
+		},
+		{
+			name: "get data from file",
+			config: &clientcmdapi.Config{
+				CurrentContext: "kubernetes",
+				Contexts:       map[string]*clientcmdapi.Context{"kubernetes": {AuthInfo: "kubernetes"}},
+				AuthInfos: map[string]*clientcmdapi.AuthInfo{"kubernetes": {
+					ClientCertificate: file.Name(),
+					ClientKey:         file.Name(),
+					TokenFile:         file.Name(),
+				},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "get data from file but file is not exist",
+			config: &clientcmdapi.Config{
+				CurrentContext: "kubernetes",
+				Contexts:       map[string]*clientcmdapi.Context{"kubernetes": {AuthInfo: "kubernetes"}},
+				AuthInfos: map[string]*clientcmdapi.AuthInfo{"kubernetes": {
+					ClientCertificate: "unknownfile",
+					ClientKey:         "unknownfile",
+					TokenFile:         "unknownfile",
+				},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "get data from config",
+			config: &clientcmdapi.Config{
+				CurrentContext: "kubernetes",
+				Contexts:       map[string]*clientcmdapi.Context{"kubernetes": {AuthInfo: "kubernetes"}},
+				AuthInfos: map[string]*clientcmdapi.AuthInfo{"kubernetes": {
+					ClientCertificateData: []byte{'f', 'o', 'o'},
+					ClientKeyData:         []byte{'b', 'a', 'r'},
+					Token:                 "k8s",
+				},
+				},
+			},
+			wantErr: false,
+		},
+	}
+	for _, rt := range tests {
+		t.Run(rt.name, func(t *testing.T) {
+			if err := EnsureAuthenticationInfoAreEmbedded(rt.config); (err != nil) != rt.wantErr {
+				t.Errorf("EnsureAuthenticationInfoAreEmbedded() error = %v, wantErr %v", err, rt.wantErr)
+			}
+		})
+	}
+}
+
+func TestEnsureCertificateAuthorityIsEmbedded(t *testing.T) {
+	file, dir, err := createtempfile(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer os.RemoveAll(dir)
+	defer file.Close()
+
+	tests := []struct {
+		name    string
+		cluster *clientcmdapi.Cluster
+		wantErr bool
+	}{
+		{
+			name:    "cluster is nil",
+			cluster: nil,
+			wantErr: true,
+		},
+		{
+			name: "get data from file",
+			cluster: &clientcmdapi.Cluster{
+				CertificateAuthority: file.Name(),
+			},
+			wantErr: false,
+		},
+		{
+			name: "get data from file but file is not exist",
+			cluster: &clientcmdapi.Cluster{
+				CertificateAuthority: "unknownfile",
+			},
+			wantErr: true,
+		},
+		{
+			name: "get data from config",
+			cluster: &clientcmdapi.Cluster{
+				CertificateAuthorityData: []byte{'f', 'o', 'o'},
+			},
+			wantErr: false,
```

Comment:
maybe organize the test cases as follows:

wantErr: false
wantErr: false
....
wantErr: true
...

first, those without error, then those with error